### ISSUE TITLE
Can't install latest dotnet-sdk in Windows Subsystem for Linux

### ISSUE THREAD

**Issue by adamtuliper** [dotnet/core-setup#4247]
## Steps to reproduce
1. Install older .NET SDK (in my case I have a working 2.1.0-rc1 though I believe we've seen similar issues with older versions prior to 2.1 and trying to upgrade those as well. Those issues only resolved with using aptitude, though I haven't done that yet on this system)

2. Try to install the latest as per: https://www.microsoft.com/net/learn/get-started/linux/ubuntu16-04

wget -q https://packages.microsoft.com/config/ubuntu/16.04/packages-microsoft-prod.deb
sudo dpkg -i packages-microsoft-prod.deb
sudo apt-get install apt-transport-https
sudo apt-get update
sudo apt-get install dotnet-sdk-2.1

## Expected  behavior
2.1 should install :)

## Actual behavior
Fails on the install dotnet-sdk-2.1 command

```
adam@home:/mnt/c/Users/Adam$ sudo apt-get install dotnet-runtime-2.1
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 dotnet-runtime-2.1 : Depends: dotnet-runtime-deps-2.1 (>= 2.1.0) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

## Environment data
Windows 10 Pro - 17134  (and prior)
Ubuntu 16.04 Windows subsystem that seemingly has been working ok 

dotnet --info returns the following

`.NET Core SDK (reflecting any global.json):
 Version:   2.1.300-rc1-008673
 Commit:    f5e3ddbe73

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  16.04
 OS Platform: Linux
 RID:         ubuntu.16.04-x64
 Base Path:   /usr/share/dotnet/sdk/2.1.300-rc1-008673/

Host (useful for support):
  Version: 2.1.0-rc1
  Commit:  eb9bc92051

.NET Core SDKs installed:
  1.0.4 [/usr/share/dotnet/sdk]
  2.1.4 [/usr/share/dotnet/sdk]
  2.1.300-rc1-008673 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0-rc1-final [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0-rc1-final [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.5 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.5 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0-preview1-26216-03 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0-rc1 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
adam@home:/mnt/c/Users/Adam$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.1.300-rc1-008673
 Commit:    f5e3ddbe73

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  16.04
 OS Platform: Linux
 RID:         ubuntu.16.04-x64
 Base Path:   /usr/share/dotnet/sdk/2.1.300-rc1-008673/

Host (useful for support):
  Version: 2.1.0-rc1
  Commit:  eb9bc92051

.NET Core SDKs installed:
  1.0.4 [/usr/share/dotnet/sdk]
  2.1.4 [/usr/share/dotnet/sdk]
  2.1.300-rc1-008673 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0-rc1-final [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0-rc1-final [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.5 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.5 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0-preview1-26216-03 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0-rc1 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

`

**Comment by leecow** [comment]
I confirmed this works OK on a clean Ubuntu WSL setup. We did make an update to package dependencies between RC and the final release so I'm wondering if that's causing this issue? You can `sudo apt autoremove aspnetcore-runtime-2.1.0-final` to clean out the RC runtime pieces or go all the way to the top and remove from the SDK with `sudo apt autoremove dotnet-sdk-2.1.300-rc1-008673`. Then, install the final 2.1 release.

**Comment by adamtuliper** [comment]
After running
sudo apt autoremove aspnetcore-runtime-2.1.0-rc1-final
Same issue. 
dotnet --info shows

`.NET Command Line Tools (1.0.4)

Product Information:
 Version:            1.0.4
 Commit SHA-1 hash:  af1e6684fd

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  16.04
 OS Platform: Linux
 RID:         ubuntu.16.04-x64
 Base Path:   /usr/share/dotnet/sdk/1.0.4

Microsoft .NET Core Shared Framework Host

  Version  : 2.1.0-preview1-26216-03
  Build    : f2c3216183d20416568a4bbf5bb7d153e826f153
`

So only 1.0.4 is left but I get the same problem as before. We've had some weird issues on WSL upgrading.  A coworker in this exact same scenario had success with aptitude (apt-get install aptitude) though last time she ran into a problem I am. I'll update with more details once I get to chat with her. 

I tried it with aptitude, note that you have to select NO for it to continue with install (we're so use to saying 'yes' for installs)

`adam@bookofadam:/usr/share/dotnet/sdk$ sudo aptitude install dotnet-sdk-2.1
The following NEW packages will be installed:
  aspnetcore-runtime-2.1{a} dotnet-hostfxr-2.1{a} dotnet-runtime-2.1{a} dotnet-runtime-deps-2.1{ab} dotnet-sdk-2.1
0 packages upgraded, 5 newly installed, 0 to remove and 42 not upgraded.
Need to get 130 MB of archives. After unpacking 375 MB will be used.
The following packages have unmet dependencies:
 dotnet-runtime-deps-2.1 : Depends: libicu52 which is a virtual package and is not provided by any available package.

The following actions will resolve these dependencies:

     Keep the following packages at their current version:
1)     aspnetcore-runtime-2.1 [Not Installed]
2)     dotnet-runtime-2.1 [Not Installed]
3)     dotnet-runtime-deps-2.1 [Not Installed]
4)     dotnet-sdk-2.1 [Not Installed]



**Accept this solution? [Y/n/q/?] n**
The following actions will resolve these dependencies:

     Install the following packages:
1)     dotnet-runtime-deps-2.1 [2.1.0-1 (xenial)]



Accept this solution? [Y/n/q/?] Y
The following NEW packages will be installed:
  aspnetcore-runtime-2.1{a} dotnet-hostfxr-2.1{a} dotnet-runtime-2.1{a} dotnet-runtime-deps-2.1{a} dotnet-sdk-2.1
0 packages upgraded, 5 newly installed, 0 to remove and 42 not upgraded.
Need to get 130 MB of archives. After unpacking 375 MB will be used.
Do you want to continue? [Y/n/?] Y
Get: 1 https://packages.microsoft.com/ubuntu/16.04/prod xenial/main amd64 dotnet-runtime-deps-2.1 amd64 2.1.0-1 [2,600 B]
Get: 2 https://packages.microsoft.com/repos/microsoft-ubuntu-trusty-prod trusty/main amd64 dotnet-hostfxr-2.1 amd64 2.1.0-1 [143 kB]
Get: 3 https://packages.microsoft.com/repos/microsoft-ubuntu-trusty-prod trusty/main amd64 dotnet-runtime-2.1 amd64 2.1.0-1 [20.3 MB]
Get: 4 https://packages.microsoft.com/repos/microsoft-ubuntu-trusty-prod trusty/main amd64 aspnetcore-runtime-2.1 amd64 2.1.0-1 [22.1 MB]
Get: 5 https://packages.microsoft.com/repos/microsoft-ubuntu-trusty-prod trusty/main amd64 dotnet-sdk-2.1 amd64 2.1.300-1 [87.5 MB]
Fetched 130 MB in 1min 22s (1,578 kB/s)
Selecting previously unselected package dotnet-runtime-deps-2.1.
(Reading database ... 40576 files and directories currently installed.)
Preparing to unpack .../dotnet-runtime-deps-2.1_2.1.0-1_amd64.deb ...
Unpacking dotnet-runtime-deps-2.1 (2.1.0-1) ...
Selecting previously unselected package dotnet-hostfxr-2.1.
Preparing to unpack .../dotnet-hostfxr-2.1_2.1.0-1_amd64.deb ...
Unpacking dotnet-hostfxr-2.1 (2.1.0-1) ...
Selecting previously unselected package dotnet-runtime-2.1.
Preparing to unpack .../dotnet-runtime-2.1_2.1.0-1_amd64.deb ...
Unpacking dotnet-runtime-2.1 (2.1.0-1) ...
Selecting previously unselected package aspnetcore-runtime-2.1.
Preparing to unpack .../aspnetcore-runtime-2.1_2.1.0-1_amd64.deb ...
Unpacking aspnetcore-runtime-2.1 (2.1.0-1) ...
        Selecting previously unselected package dotnet-sdk-2.1.
Preparing to unpack .../dotnet-sdk-2.1_2.1.300-1_amd64.deb ...
Unpacking dotnet-sdk-2.1 (2.1.300-1) ...
Processing triggers for man-db (2.7.5-1) ...
Setting up dotnet-runtime-deps-2.1 (2.1.0-1) ...
Setting up dotnet-hostfxr-2.1 (2.1.0-1) ...
Setting up dotnet-runtime-2.1 (2.1.0-1) ...
Setting up aspnetcore-runtime-2.1 (2.1.0-1) ...
Setting up dotnet-sdk-2.1 (2.1.300-1) ...
This software may collect information about you and your use of the software, and send that to Microsoft.
Please visit http://aka.ms/dotnet-cli-eula for more information.
Welcome to .NET Core!
---------------------
`

**Comment by adamtuliper** [comment]
For the life of me can't get those edits above to show up right, hopefully they show better for you than me :)

**Comment by jackiebecker** [comment]
I am having this same issue--I have rc1 installed (using aptitude because apt-get failed) and am now attempting to install 2.1 using `sudo apt-get install dotnet-sdk-2.1`. This errored out with the same message on both my laptop (Surface) and my personal desktop (also Windows) on Ubuntu/bash. As noted above, worked fine using aptitude, a solution proposed in this post by someone having the same problem: https://askubuntu.com/questions/363200/e-unable-to-correct-problems-you-have-held-broken-packages/451078. While this is a functional workaround, it would be nice to know what is going on here as it seems like the problem is occurring for a number of WSL users.

**Comment by adamtuliper** [comment]
@leecow Any thoughts on how we can get to the root issue here?

**Comment by leecow** [comment]
I have seen various apt clients deal with dependency resolutions differently though I can't say why that is. Will find Rich Turner's handle and add him to this thread to see if there's anything on the WSL side that could be thought about. 

While it's a bit of a hammer, completely cleaning .NET Core out of the system then reinstalling the needed versions has gotten me out of any weird state I've managed to get myself into.  

* `sudo apt-get autoremove dotnet-host` starts at the bottom of the dependency tree rather than the top and will do most of the work.  
* `sudo rm -rf /usr/share/dotnet/` to clear up the remaining things in the sdk/ directory. 
* reinstall the release versions you need to work with. Remember that newer versions of the SDK can be used to build for older versions of the runtime so latest released SDK (2.1.301 at the moment) should be all that you need.

**Comment by jefking** [comment]
I am also having this issue. I don't believe I have ever had dotnet installed in WSL on this laptop. Windows is up-to-date.
Version:
```
lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.4 LTS
Release:        16.04
Codename:       xenial
```
Run dotnet:
```
dotnet: command not found
```
Installation (Like Above):
```
root@DESKTOP-74N9G6D:/mnt/c/Users/jking/Documents/GitHub# wget -q https://packages.microsoft.com/config/ubuntu/16.04/packages-microsoft-prod.deb
root@DESKTOP-74N9G6D:/mnt/c/Users/jking/Documents/GitHub# sudo dpkg -i packages-microsoft-prod.deb
(Reading database ... 36796 files and directories currently installed.)
Preparing to unpack packages-microsoft-prod.deb ...
Unpacking packages-microsoft-prod (1.0-3) over (1.0-3) ...
Setting up packages-microsoft-prod (1.0-3) ...
root@DESKTOP-74N9G6D:/mnt/c/Users/jking/Documents/GitHub# sudo apt-get install apt-transport-https
Reading package lists... Done
Building dependency tree
Reading state information... Done
apt-transport-https is already the newest version (1.2.26).
0 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.
root@DESKTOP-74N9G6D:/mnt/c/Users/jking/Documents/GitHub# sudo apt-get update
Get:1 http://security.ubuntu.com/ubuntu xenial-security InRelease [107 kB]
Hit:2 http://archive.ubuntu.com/ubuntu xenial InRelease
Get:3 http://archive.ubuntu.com/ubuntu xenial-updates InRelease [109 kB]
Hit:4 https://deb.nodesource.com/node_8.x xenial InRelease
Hit:5 https://packages.microsoft.com/repos/microsoft-ubuntu-trusty-prod trusty InRelease
Hit:6 https://packages.microsoft.com/ubuntu/16.04/prod xenial InRelease
Fetched 216 kB in 5s (36.0 kB/s)
Reading package lists... Done
root@DESKTOP-74N9G6D:/mnt/c/Users/jking/Documents/GitHub# sudo apt-get install dotnet-sdk-2.1
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 dotnet-sdk-2.1 : Depends: dotnet-runtime-2.1 (>= 2.1.1) but it is not going to be installed
                  Depends: aspnetcore-runtime-2.1 (>= 2.1.1) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```
Running up the dependency chain:
```
root@DESKTOP-74N9G6D:/mnt/c/Users/jking/Documents/GitHub# sudo apt-get install dotnet-runtime-2.1
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 dotnet-runtime-2.1 : Depends: dotnet-runtime-deps-2.1 (>= 2.1.1) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
root@DESKTOP-74N9G6D:/mnt/c/Users/jking/Documents/GitHub# sudo apt-get install dotnet-runtime-deps-2.1
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 dotnet-runtime-deps-2.1 : Depends: libicu52 but it is not installable
E: Unable to correct problems, you have held broken packages.
root@DESKTOP-74N9G6D:/mnt/c/Users/jking/Documents/GitHub# sudo apt-get install libicu52
Reading package lists... Done
Building dependency tree
Reading state information... Done
Package libicu52 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'libicu52' has no installation candidate
```

**Comment by jefking** [comment]
Found this: [No installation candidate for “libicu52” on 15.10](https://askubuntu.com/questions/722075/no-installation-candidate-for-libicu52-on-15-10)

Ran
```
wget http://mirrors.kernel.org/ubuntu/pool/main/i/icu/libicu52_52.1-3ubuntu0.8_amd64.deb
dpkg -i libicu52*.deb
```
Then
```
sudo apt-get install dotnet-sdk-2.1
```
dotnet
```
root@DESKTOP-74N9G6D:~# dotnet

Usage: dotnet [options]
Usage: dotnet [path-to-application]

Options:
  -h|--help         Display help.
  --info            Display .NET Core information.
  --list-sdks       Display the installed SDKs.
  --list-runtimes   Display the installed runtimes.

path-to-application:
  The path to an application .dll file to execute.
```

**Comment by blackrek** [comment]
fixed for ubuntu server 18.04 for dotnet-sdk-2.1

1. remove old version (thx @leecow comments):
> `sudo apt-get autoremove dotnet-host` starts at the bottom of the dependency tree rather than the top and will do most of the work.
`sudo rm -rf /usr/share/dotnet/` to clear up the remaining things in the sdk/ directory.

2. fix @jefking instructions (replace package version for libicu60):
`wget http://mirrors.kernel.org/ubuntu/pool/main/i/icu/libicu60_60.2-6ubuntu1_amd64.deb
dpkg -i libicu60_60.2-6ubuntu1_amd64.deb`

3. now `apt-get install dotnet-sdk-2.1` works!

today installed 2.1.302 version
thx all!

**Comment by leecow** [comment]
Awesome to see some successes in this thread :-). @jackiebecker, have you been able to get up and running with 2.1? If not, have you tried a complete clean up as described above (https://github.com/dotnet/core-setup/issues/4247#issuecomment-400686996)?

**Comment by EdenRidgway** [comment]
To get it to install on Ubuntu 18.04 I had to run:

**Comment by mytee306** [comment]
> fixed for ubuntu server 18.04 for dotnet-sdk-2.1
> 
> 1. remove old version (thx @leecow comments):
> 
> > `sudo apt-get autoremove dotnet-host` starts at the bottom of the dependency tree rather than the top and will do most of the work.
> > `sudo rm -rf /usr/share/dotnet/` to clear up the remaining things in the sdk/ directory.
> 
> 1. fix @jefking instructions (replace package version for libicu60):
>    `wget http://mirrors.kernel.org/ubuntu/pool/main/i/icu/libicu60_60.2-6ubuntu1_amd64.deb dpkg -i libicu60_60.2-6ubuntu1_amd64.deb`
> 
> (if you need other version of libicu, just found it in [http://mirrors.edge.kernel.org/ubuntu/pool/main/i/icu/](url))
> 
> 1. now `apt-get install dotnet-sdk-2.1` works!
> 
> today installed 2.1.302 version
> thx all!

It got stuck on the line
Connecting to 2.0 (2.0)|2.0.0.0|:80... 
Unsuccessfully retrying

**Comment by blackrek** [comment]
at what step did the error occur?

**Comment by mytee306** [comment]
> at what step did the error occur?

Thank you for asking @blackrek 

1. sudo apt-get autoremove dotnet-host

2. sudo rm -rf /usr/share/dotnet/

3. wget http://mirrors.kernel.org/ubuntu/pool/main/i/icu/libicu60_60.2-6ubuntu1_amd64.deb dpkg -i libicu60_60.2-6ubuntu1_amd64.deb

Also http://mirrors.edge.kernel.org/ubuntu/pool/main/i/icu/ is not found

**Comment by blackrek** [comment]
Try proxy.
I have access to this resources

**Comment by blitzblade** [comment]
@EdenRidgway thanks! your solution worked for me.
sudo wget http://mirrors.kernel.org/ubuntu/pool/main/i/icu/libicu55_55.1-7_amd64.deb
 
Actually this version is what solves the issue for ubuntu 18.

**Comment by mytee306** [comment]
> Try proxy.
> I have access to this resources

I've tried installing libicu60_60.2-6ubuntu1_i386.deb
Which I got from from http://mirrors.edge.kernel.org/ubuntu/pool/main/i/icu/ 
And I got the following: Error: Dependency is not satisfiable: libgcc1 (>= 1:7)

**Comment by clearhat** [comment]
This helped me get installed to 2.1.3, but I'm back again because I'm trying to upgrade to 2.1.4, and getting nowhere. All instructions I've found says to use 

`apt-get install dotnet-sdk-2.1`

This installs 2.1.3 currently. So how do I upgrade to 2.1.4 (or 2.1.5 for that matter, which just came out?)

**Comment by richstep** [comment]
> > at what step did the error occur?
> 
> Thank you for asking @blackrek
> 
> 1. sudo apt-get autoremove dotnet-host
> 2. sudo rm -rf /usr/share/dotnet/
> 3. wget http://mirrors.kernel.org/ubuntu/pool/main/i/icu/libicu60_60.2-6ubuntu1_amd64.deb dpkg -i libicu60_60.2-6ubuntu1_amd64.deb
> 
> Also http://mirrors.edge.kernel.org/ubuntu/pool/main/i/icu/ is not found

This answer for me, but I had to break step 3 into to parts:
3a.  wget http://mirrors.kernel.org/ubuntu/pool/main/i/icu/libicu60_60.2-6ubuntu1_amd64.deb
3b. sudo dpkg -i libicu60_60.2-6ubuntu1_amd64.deb

4. sudo apt-get install dotnet-sdk-2.1
Output:
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following additional packages will be installed:
  aspnetcore-runtime-2.1 dotnet-host dotnet-hostfxr-2.1 dotnet-runtime-2.1 dotnet-runtime-deps-2.1 liblttng-ust-ctl2
  liblttng-ust0 liburcu4
The following NEW packages will be installed:
  aspnetcore-runtime-2.1 dotnet-host dotnet-hostfxr-2.1 dotnet-runtime-2.1 dotnet-runtime-deps-2.1 dotnet-sdk-2.1
  liblttng-ust-ctl2 liblttng-ust0 liburcu4
0 upgraded, 9 newly installed, 0 to remove and 107 not upgraded.
Need to get 131 MB of archives.
After this operation, 378 MB of additional disk space will be used.
Do you want to continue? [Y/n] Y
.....
Welcome to .NET Core!
---------------------

**Comment by maitreya15** [comment]
helpful...Thanks!!

**Comment by tripolskypetr** [comment]
Just install this .deb package
https://packages.ubuntu.com/ru/xenial/libicu55

**Comment by CoskunSunali** [comment]
Are not these additional steps better to be a part of the documentation at https://dotnet.microsoft.com/download/linux-package-manager/ubuntu18-04/sdk-2.2.105 or even better set as dependencies so `apt` or `apt-get` installs them automatically?

It is 2019 and we have to search through Google to figure out how to install dotnet-sdk using a package manager. Feels sad.

**Comment by kwienq101** [comment]
I also had the same error as CoskunSunali. 
...

**Comment by Aliaksandr3by** [comment]
dotnet-sdk-2.2

https://ubuntu.pkgs.org/18.10/ubuntu-main-amd64/libicu60_60.2-6ubuntu1_amd64.deb.html
https://debian.pkgs.org/8/debian-main-amd64/libssl1.0.0_1.0.1t-1+deb8u8_amd64.deb.html
sudo dpkg -i libicu60_60.2-6ubuntu1_amd64.deb
sudo dpkg -i libssl1.0.0_1.0.1t-1+deb8u8_amd64.deb

sudo apt install dotnet-runtime-deps-2.2
sudo apt-get install dotnet-sdk-2.2

**Comment by dagood** [comment]
> Are not these additional steps [...] better set as dependencies so `apt` or `apt-get` installs them automatically?

They are dependencies, there's just something going wrong when apt tries to find them.

There isn't quite enough info in these posts to know for sure it's what's going on, but it seems like the issue is that the `packages-microsoft-prod.deb` installed on the machine doesn't match the current Ubuntu release installed. For example, running Ubuntu 18.04 and trying to install dotnet, but having the 16.04 `packages-microsoft-prod.deb` installed.

Here are some extra diagnosis steps to know for sure, and I think they'll still give some useful info even if you've already worked around the issue by grabbing a `libicu` deb file from online:

```sh
# Shows the repositories you have configured. Source: https://askubuntu.com/a/741948
grep -r --include '*.list' '^deb ' /etc/apt/sources.list /etc/apt/sources.list.d/
# Show what version of libicu you have available to you.
apt search libicu
# Show your distro version.
cat /etc/os-release
# This install command will show exactly which dependency can't be met.
sudo apt-get install dotnet-runtime-deps-2.1
```

A good list of repositories looks something like this. `xenial` for the Ubuntu repositories matches `xenial` for the Microsoft package repository:

```
[...]
/etc/apt/sources.list:deb http://security.ubuntu.com/ubuntu/ xenial-security multiverse
/etc/apt/sources.list.d/microsoft-prod.list:deb [arch=amd64]
  https://packages.microsoft.com/ubuntu/16.04/prod xenial main
```

A bad list of repositories looks like this, mixing `bionic` with `xenial`:

```
/etc/apt/sources.list:deb http://security.ubuntu.com/ubuntu/ bionic-security multiverse
/etc/apt/sources.list.d/microsoft-prod.list:deb [arch=amd64]
  https://packages.microsoft.com/ubuntu/16.04/prod xenial main
```

The reason the mismatch between the Ubuntu repositories and the Microsoft repository causes problems is that the ICU lib package has a different name (not just version!) for each Ubuntu release. This means we need a new Microsoft feed with a unique `dotnet-runtime-deps-*` package for every version.

For example, if you have Ubuntu **18.04** installed, you are *only* able to download `libicu60` from the Ubuntu repository. If you try to install `dotnet-runtime-deps-2.1` from the **16.04** Microsoft repository, it will fail, because it depends on `libicu55`, only available in Ubuntu **16.04**.

One way you can get into this situation without realizing it is to install dotnet on Ubuntu 16.04, then do a release upgrade to Ubuntu 18.04. The Microsoft feed package is stuck at 16.04--it doesn't have any way to update to match the change in distro release version.

The reason installing a `libicu##` from the web works is that it bypasses the Ubuntu package repository, and you can get any version you want. That way, you can force a 16.04 `dotnet-runtime-deps-*` to work on Ubuntu 18.04.

A fix that brings you back to more standard behavior is to make sure the correct Microsoft package repository deb is installed:

```sh
# This is copied from the 18.04 instructions, make sure to get the one matching your release!
wget -q https://packages.microsoft.com/config/ubuntu/18.04/packages-microsoft-prod.deb
sudo dpkg -i packages-microsoft-prod.deb
```

(Side note: something that I ran into while trying this out is that `wget` will create `packages-microsoft-prod.deb.1` if `packages-microsoft-prod.deb` already exists. This means `sudo dpkg -i packages-microsoft-prod.deb` installed the package I downloaded earlier. This is kind of a silly thing to mention, but it threw me off for a bit so I thought I'd include it.)

---

I'm closing this issue, but please do let me know if the above doesn't work or the diagnosis steps give you weird looking results.

There's nothing I know of that we can do in Core-Setup to improve this based on my current understanding. If someone knows some Deb package tricks (or things specific to Ubuntu), I welcome suggestions.

**Comment by cwash** [comment]
Looks like some package wiring was missing, but specifying a few transitive deps explicitly seemed to work for me...

> #!/bin/bash
> wget https://packages.microsoft.com/config/ubuntu/18.04/packages-microsoft-prod.deb
> wget http://mirrors.kernel.org/ubuntu/pool/main/i/icu/libicu55_55.1-7_amd64.deb
> sudo dpkg -i packages-microsoft-prod.deb
> sudo dpkg -i libicu55_55.1-7_amd64.deb
> sudo add-apt-repository universe
> sudo apt-get install -y apt-transport-https
> sudo apt-get update
> sudo apt-get install -y dotnet-sdk-2.1 dotnet-runtime-2.1 aspnetcore-runtime-2.1 dotnet-runtime-deps-2.1 libicu55
